File: gin/library/casbin.go
package library

import (
	"fmt"

	"github.com/casbin/casbin/v2"
	_ "github.com/go-sql-driver/mysql"
)

var CasBinDB *casbin.Enforcer

// CasBin :权限控制
type CasBin struct{}

// CasBinPool: 连接
func CasBinPool() {
	e, err := casbin.NewEnforcer("config/casbin.conf", "config/casbin.csv")
	if err != nil {
		fmt.Println("[Casbin] Open:", err)
		return
	}
	CasBinDB = e
}

// Verify :验证
func (c *CasBin) Verify(sub string, obj string, act string) bool {
	if CasBinDB == nil {
		return false
	}
	res, err := CasBinDB.Enforce(sub, obj, act)
	fmt.Println(sub, obj, act)
	if err != nil {
		fmt.Println("[Casbin] Verify:", err)
		return false
	}
	return res
}

// Add :添加
func (c *CasBin) Add(sub string, obj string, act string) bool {
	if CasBinDB == nil {
		return false
	}
	res, err := CasBinDB.AddPolicy(sub, obj, act)
	if err != nil {
		fmt.Println("[Casbin] Add:", err)
	}
	if err := CasBinDB.SavePolicy(); err != nil {
		fmt.Println("[Casbin] Save:", err)
	}
	return res
}

// Add :删除
func (c *CasBin) Remove(sub string, obj string, act string) bool {
	if CasBinDB == nil {
		return false
	}
	res, err := CasBinDB.RemovePolicy(sub, obj, act)
	if err != nil {
		fmt.Println("[Casbin] Remove:", err)
	}
	if err := CasBinDB.SavePolicy(); err != nil {
		fmt.Println("[Casbin] Save:", err)
	}
	return res
}

File: java/src/main/java/webmis/library/Casbin.java
package webmis.library;

import org.casbin.jcasbin.main.Enforcer;
import webmis.base.Base;

/* 权限控制 */
public class Casbin extends Base {

  public static Enforcer CasBinDB;    //连接

  /* 连接 */
  public void CasBinPool() {
    try {
      CasBinDB = new Enforcer("config/casbin.conf", "config/casbin.csv");
    } catch (Exception e) {
      Print("[Casbin] Conn:", e.getMessage());
    }
  }

  /* 验证 */
  public boolean Verify(String sub, String obj, String act) {
    if(CasBinDB==null) return false;
    return CasBinDB.enforce(sub, obj, act);
  }

  /* 添加 */
  public void Add(String sub, String obj, String act) {
    Print("conn", CasBinDB);
    if(CasBinDB==null) return ;
    boolean res = CasBinDB.addPolicy(sub, obj, act);
    Print(res);
    try {
      CasBinDB.savePolicy();
    } catch (Exception e) {
      Print("[Casbin] Add:", e.getMessage());
    }
    // if(CasBinDB.addPolicy(sub, obj, act)) CasBinDB.savePolicy();
    // CasBinDB.savePolicy();
  }

  /* 删除 */
  public void Remove(String sub, String obj, String act) {
    if(CasBinDB==null) return ;
    CasBinDB.removePolicy(sub, obj, act);
    CasBinDB.savePolicy();
  }
  
}
